src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+            "client_secret": client_secret,
+            "grant_type": "authorization_code",
+        }
+
+        body = urllib_parse.urlencode(args)
+        http_client = httpclient.AsyncHTTPClient()
+        response = await http_client.fetch(
+            self._OAUTH_ACCESS_TOKEN_URL % self._GITLAB_PREFIX,
+            method='POST',
+            headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            body=body,
+            raise_error=False)
+
+        response_values = {}
+        if response.body:
+            response_values = escape.json_decode(response.body)
+
+        if response.error:
+            if response_values.get('error_description'):
+                error_text = response_values.get('error_description')
+            elif response_values.get('error'):
+                error_text = response_values.get('error')
+            else:
+                error_text = str(response.error)
+
+            error_message = 'Failed to load access_token: ' + error_text
+            LOGGER.error(error_message)
+            raise AuthFailureError(error_message)
+
+        access_token = response_values.get('access_token')
+
+        if not access_token:
+            message = 'No access token in response: ' + str(response.body)
+            LOGGER.error(message)
+            raise AuthFailureError(message)
+
+        user = await self.fetch_user(access_token)
+
+        if user is None:
+            error_message = 'Failed to load user info'
+            LOGGER.error(error_message)
+            raise AuthFailureError(error_message)
+
+        return {**response_values, **user}
+
+    async def fetch_user(self, access_token):
+        user = await self.oauth2_request(
+            self._OAUTH_GITLAB_USERINFO % self._GITLAB_PREFIX,
+            access_token)
+        if user is None:
+            return None
+
+        fieldmap = {}
+        for field in {"id", "username", "name", "email", "state"}:
+            fieldmap[field] = user.get(field)
+
+        return fieldmap
+
+# noinspection PyProtectedMember
+class GitlabOAuthAuthenticator(auth_base.Authenticator, GitlabOAuth2Mixin):
+    def __init__(self, params_dict):
+        super().__init__()
+
+        LOGGER.debug("Init gitlab oauth provider with " + str(params_dict))
+
+        self.client_id = model_helper.read_obligatory(params_dict, 'client_id', ' for Gitlab OAuth')
+
+        secret_value = model_helper.read_obligatory(params_dict, 'secret', ' for Gitlab OAuth')
+        self.secret = model_helper.resolve_env_vars(secret_value, full_match=True)
+
+        gitlabPrefix = params_dict.get('url')
+        if not model_helper.is_empty(gitlabPrefix):
+            self._GITLAB_PREFIX = gitlabPrefix
+
+        self.states = {}
+        self.user_states = {}
+        self.gitlab_update = params_dict.get('auth_info_ttl')
+        self.gitlab_dump = params_dict.get('state_dump_file')
+        self.gitlab_group_support = params_dict.get('group_support', True)
```

Comment:
Do you think there are use cases when we don't want to load GitLab groups?

Reply:
As I described in doc, there is big difference in oauth scopes. Scope - it's type of access to different endpoints in API.

Safest scope is "read_user" - it's readonly access to only one endpoint -user info. But this scope is not enough to get group list. 

Second scope is `api` - its read/write access to all API. It's allow to get groups but in case of token leak - gives much more access to Gitlab.

So I decided to give end user the choice - groups or safe token.

Oh, just found that new scope `read_api` introduced in fresh version. Probably need new option :)